File: tests/domain/observation/freshness.test.ts
import {
  freshnessOf,
  ANCIENNE_APRES_MS,
  PERIMEE_APRES_MS,
  type Freshness,
} from "@domain/observation/freshness";

const MAINTENANT = new Date("2026-08-01T12:00:00Z");
const HEURE = 60 * 60 * 1000;

/** Observation datée de `ms` millisecondes avant `MAINTENANT`. */
const ilYA = (ms: number): Date => new Date(MAINTENANT.getTime() - ms);

/**
 * BR-005 fixe des seuils ABSOLUS : 2 h puis 24 h. Ce ne sont pas des multiples
 * du TTL de cache — 03-conception.md § 4.1 parle de « 2 × TTL », mais il y
 * décrit la fraîcheur du CACHE, pas l'âge de la MESURE. Les confondre
 * déclarerait périmée une observation de 40 minutes.
 *
 * L'âge se calcule sur `date_obs`, jamais sur la date de récupération (BR-005).
 */
describe("fraîcheur d'une observation hydrométrique (BR-005)", () => {
  it("expose les seuils de la règle, pas des nombres magiques", () => {
    expect(ANCIENNE_APRES_MS).toBe(2 * HEURE);
    expect(PERIMEE_APRES_MS).toBe(24 * HEURE);
  });

  // Les quatre cas que BR-005 § « Vérifiable par » prescrit nommément.
  it("1 h 59 → fraîche", () => {
    expect(freshnessOf(ilYA(1 * HEURE + 59 * 60_000), MAINTENANT)).toBe<Freshness>("Fraiche");
  });

  it("2 h 01 → ancienne", () => {
    expect(freshnessOf(ilYA(2 * HEURE + 60_000), MAINTENANT)).toBe<Freshness>("Ancienne");
  });

  it("23 h 59 → ancienne", () => {
    expect(freshnessOf(ilYA(23 * HEURE + 59 * 60_000), MAINTENANT)).toBe<Freshness>("Ancienne");
  });

  it("24 h 01 → périmée", () => {
    expect(freshnessOf(ilYA(24 * HEURE + 60_000), MAINTENANT)).toBe<Freshness>("Perimee");
  });

  it("bascule exactement AU seuil, pas après", () => {
    // La borne appartient à l'état le plus sévère : on ne minimise jamais l'âge.
    expect(freshnessOf(ilYA(2 * HEURE), MAINTENANT)).toBe<Freshness>("Ancienne");
    expect(freshnessOf(ilYA(24 * HEURE), MAINTENANT)).toBe<Freshness>("Perimee");
  });

  it("traite une date future comme fraîche plutôt que d'échouer", () => {
    // Une horloge d'appareil en avance ne doit pas casser l'affichage d'une
    // carte. Un âge négatif reste un âge inférieur au seuil.
    expect(freshnessOf(new Date(MAINTENANT.getTime() + 60_000), MAINTENANT)).toBe<Freshness>(
      "Fraiche",
    );
  });

  it("gère l'écart réel constaté en production", () => {
    // Le 2026-07-30, l'âge des observations allait de 7 minutes à 9 jours selon
    // la station (BR-001). Les deux extrêmes doivent être classés correctement.
    expect(freshnessOf(ilYA(7 * 60_000), MAINTENANT)).toBe<Freshness>("Fraiche");
    expect(freshnessOf(ilYA(9 * 24 * HEURE), MAINTENANT)).toBe<Freshness>("Perimee");
  });
});

describe("date de mesure inexploitable", () => {
  it("ne présente jamais une date invalide comme fraîche", () => {
    // NaN >= seuil est faux deux fois : sans garde, la fonction retombe sur
    // « Fraiche », soit l'état le MOINS sévère pour une donnée dont on ne sait
    // rien. C'est l'inverse de la règle que porte cette fonction.
    expect(freshnessOf(new Date(""), MAINTENANT)).toBe<Freshness>("Perimee");
    expect(freshnessOf(new Date("pas-une-date"), MAINTENANT)).toBe<Freshness>("Perimee");
  });

  it("ne présente pas non plus un instant courant invalide comme fraîche", () => {
    expect(freshnessOf(ilYA(1000), new Date("n'importe quoi"))).toBe<Freshness>("Perimee");
  });
});

File: test/domain/observation/freshness_test.dart
// Verrouille les bornes absolues de BR-005 (2 h puis 24 h), sur la date de
// MESURE et jamais de récupération (BR-001). `now` est un paramètre : sans
// lui, les bornes ne sont pas testables.
import 'package:flutter_test/flutter_test.dart';
import 'package:martinpecheur/domain/observation/freshness.dart';

void main() {
  group('Fraîcheur d\'une observation (BR-005)', () {
    test('un âge de 1 min ou de 1 h 59 est fraiche', () {
      final DateTime now = DateTime.utc(2026, 9, 13, 12);

      expect(
        freshnessOf(
          measuredAt: now.subtract(const Duration(minutes: 1)),
          now: now,
        ),
        Freshness.fraiche,
      );
      expect(
        freshnessOf(
          measuredAt: now.subtract(const Duration(hours: 1, minutes: 59)),
          now: now,
        ),
        Freshness.fraiche,
      );
    });

    test('un âge de 2 h exactement est ancienne — la borne est la plus '
        'sévère', () {
      final DateTime now = DateTime.utc(2026, 9, 13, 12);

      expect(
        freshnessOf(
          measuredAt: now.subtract(const Duration(hours: 2)),
          now: now,
        ),
        Freshness.ancienne,
      );
    });

    test('un âge de 2 h 1 min ou de 23 h 59 est ancienne', () {
      final DateTime now = DateTime.utc(2026, 9, 13, 12);

      expect(
        freshnessOf(
          measuredAt: now.subtract(const Duration(hours: 2, minutes: 1)),
          now: now,
        ),
        Freshness.ancienne,
      );
      expect(
        freshnessOf(
          measuredAt: now.subtract(const Duration(hours: 23, minutes: 59)),
          now: now,
        ),
        Freshness.ancienne,
      );
    });

    test('un âge de 24 h exactement est perimee — la borne est la plus '
        'sévère', () {
      final DateTime now = DateTime.utc(2026, 9, 13, 12);

      expect(
        freshnessOf(
          measuredAt: now.subtract(const Duration(hours: 24)),
          now: now,
        ),
        Freshness.perimee,
      );
    });

    test('un âge de 24 h 1 min est perimee, comme les dix-sept jours '
        "constatés sur K447001001 le 2026-09-13", () {
      final DateTime now = DateTime.utc(2026, 9, 13, 12);

      expect(
        freshnessOf(
          measuredAt: now.subtract(const Duration(hours: 24, minutes: 1)),
          now: now,
        ),
        Freshness.perimee,
      );
      expect(
        freshnessOf(
          measuredAt: DateTime.utc(2026, 8, 27, 8),
          now: DateTime.utc(2026, 9, 13, 12),
        ),
        Freshness.perimee,
      );
    });

    test('une mesure datée dans le futur est fraiche : un âge négatif est une '
        "anomalie de la source, pas une donnée vieille", () {
      final DateTime now = DateTime.utc(2026, 9, 13, 12);

      expect(
        freshnessOf(measuredAt: now.add(const Duration(hours: 3)), now: now),
        Freshness.fraiche,
      );
    });

    test("difference compare des instants absolus : une mesure UTC à -3 h "
        'reste ancienne même comparée à un now exprimé en heure locale '
        '(pas de toUtc() "au cas où")', () {
      final DateTime measuredAtUtc = DateTime.utc(2026, 9, 13, 9);
      final DateTime nowLocal = DateTime.utc(2026, 9, 13, 12).toLocal();

      expect(
        freshnessOf(measuredAt: measuredAtUtc, now: nowLocal),
        Freshness.ancienne,
      );
    });

    test('ancienneApres vaut 2 h et perimeeApres vaut 24 h', () {
      expect(ancienneApres, const Duration(hours: 2));
      expect(perimeeApres, const Duration(hours: 24));
    });
  });
}
